pkg/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -904,9 +915,152 @@ func (q *DBJobQueue) jobDependents(ctx context.Context, conn connection, id uuid
 	return dependents, nil
 }
 
+// listJobs returns a list of all of the job UUIDs
+func (q *DBJobQueue) listJobs(ctx context.Context, conn connection) (jobs []uuid.UUID, err error) {
+	rows, err := conn.Query(ctx, sqlQueryListJobs)
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var t uuid.UUID
+		err = rows.Scan(&t)
+		if err != nil {
+			// Log the error and try to continue with the next row
+			q.logger.Error(err, "Unable to read job uuid from jobs")
+			continue
+		}
+		jobs = append(jobs, t)
+	}
+	if rows.Err() != nil {
+		q.logger.Error(rows.Err(), "Error reading job uuids from jobs")
+	}
+
+	return
+}
+
 // AllRootJobIDs returns a list of top level job UUIDs that the worker knows about
-func (q *DBJobQueue) AllRootJobIDs() ([]uuid.UUID, error) {
-	// TODO write this
+func (q *DBJobQueue) AllRootJobIDs(ctx context.Context) (rootJobs []uuid.UUID, err error) {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return
+	}
+	defer conn.Release()
 
-	return nil, nil
+	var jobs []uuid.UUID
+	jobs, err = q.listJobs(ctx, conn)
+	if err != nil {
+		return
+	}
+
+	for _, id := range jobs {
+		var dependents []uuid.UUID
+		dependents, err = q.jobDependents(ctx, conn, id)
+		if err != nil {
+			return
+		}
+
+		if len(dependents) == 0 {
+			rootJobs = append(rootJobs, id)
+		}
+	}
+
+	return
+}
+
+// DeleteJob deletes a job and all of its dependencies from the database
+// If a dependency has multiple dependents it will only remove the parent job from
+// the dependents list for that job instead of removing it.
+//
+// This assumes that the jobs have been created correctly, and that they have
+// no dependency loops. Shared Dependents are ok, but a job cannot have a dependency
+// on any of its parents (this should never happen).
+func (q *DBJobQueue) DeleteJob(ctx context.Context, id uuid.UUID) error {
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	tx, err := conn.Begin(context.Background())
```

Comment:
Few overlooks, this will carry the context over into the transaction so we can cancel it or put a deadline on it etc.

```suggestion
	tx, err := conn.Begin(ctx)
```